src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs

```diff
@@ -234,30 +234,81 @@ void ErrorHandler(object? sender, Newtonsoft.Json.Serialization.ErrorEventArgs e
+            // but ErrorContext.Path does not include the missing required property name like we want it to.
+            // For example, given the following class and input missing the required "Name" property:
+            //
+            // class Person
+            // {
+            //     [JsonProperty(Required = Required.Always)]
+            //     public string Name { get; set; }
+            // }
+            //
+            // We will see the following ErrorContext:
+            //
+            // Error   {"Required property 'Name' not found in JSON. Path 'Person'..."} System.Exception {Newtonsoft.Json.JsonSerializationException}
+            // Member  "Name"   object {string}
+            // Path    "Person" string
+            //
+            // So we update the path used for the ModelStateDictionary key to be "Person.Name" instead of just "Person".
+            // See https://github.com/aspnet/Mvc/issues/8509
             var path = eventArgs.ErrorContext.Path;
-            var member = eventArgs.ErrorContext.Member?.ToString();
-            var addMember = !string.IsNullOrEmpty(member);
+            var member = eventArgs.ErrorContext.Member as string;
+
+            // There are some deserialization exceptions that include the member in the path but not at the end.
+            // For example, given the following classes and invalid input like { "b": { "c": { "d": abc } } }:
+            //
+            // class A
+            // {
+            //     public B B { get; set; }
+            // }
+            // class B 
+            // {
+            //     public C C { get; set; }
+            // }
+            // class C 
+            // {
+            //     public string D { get; set; }
+            // }
+            //
+            // We will see the following ErrorContext:
+            //
+            // Error   {"Unexpected character encountered while parsing value: b. Path 'b.c.d'..."} System.Exception {Newtonsoft.Json.JsonReaderException}
+            // Member  "c"     object {string}
+            // Path    "b.c.d" string
+            //
+            // Notice that Member "c" is in the middle of the Path "b.c.d". The error handler gets invoked for each level of nesting.
+            // null, "b", "c" and "d" are each a Member in different ErrorContexts all reporting the same parsing error.
+            //
+            // The parsing error is reported as a JsonReaderException instead of as a JsonSerializationException like
+            // for missing required properties. We use the exception type to filter out these errors and keep the path used
+            // for the ModelStateDictionary key as "b.c.d" instead of "b.c.d.c"
+            // See https://github.com/dotnet/aspnetcore/issues/33451 
+            var addMember = !string.IsNullOrEmpty(member) && eventArgs.ErrorContext.Error is JsonSerializationException;
```

Comment:
This additional `&& eventArgs.ErrorContext.Error is JsonSerializationException` check is the key to the fix.

Reply:
Is `JsonSerializationException` the only exception type other than `JsonReaderException` that will surface in the error context? AKA, is this going to inadvertently break the way the member path is reported for other exceptions?

Reply:
This is why I dug up the original PR that added this logic. As far as I can tell, this logic was added only to better report errors for missing required properties. If this happened to improve the path for other exception types, that was an unintended side effect which there are no tests for.

This and other issues like those involving escaped keys shows that the `addMember` logic is prone to adding the member to the path when it shouldn't, so my goal was to append the member to the path in the fewest cases possible without regressing the missing required property scenario.